File: backend-java/src/main/java/com/neria/manager/usage/UsageAlertsService.java
package com.neria.manager.usage;

import com.neria.manager.common.entities.Policy;
import com.neria.manager.common.entities.Tenant;
import com.neria.manager.common.repos.PolicyRepository;
import com.neria.manager.common.repos.TenantRepository;
import java.util.ArrayList;
import java.util.List;
import org.springframework.stereotype.Service;

@Service
public class UsageAlertsService {
  private final TenantRepository tenantRepository;
  private final PolicyRepository policyRepository;
  private final UsageService usageService;

  public UsageAlertsService(
      TenantRepository tenantRepository,
      PolicyRepository policyRepository,
      UsageService usageService) {
    this.tenantRepository = tenantRepository;
    this.policyRepository = policyRepository;
    this.usageService = usageService;
  }

  public List<UsageAlert> list(String tenantId) {
    List<Tenant> tenants =
        tenantId != null
            ? tenantRepository.findById(tenantId).map(List::of).orElse(List.of())
            : tenantRepository.findAll();
    List<UsageAlert> alerts = new ArrayList<>();

    for (Tenant tenant : tenants) {
      if (tenant.isKillSwitch()) {
        alerts.add(new UsageAlert(tenant.getId(), "kill_switch", "critical", "Kill switch activo", null, null));
      }
      Policy policy = policyRepository.findByTenantId(tenant.getId()).orElse(null);
      if (policy == null) {
        continue;
      }
      UsageService.Totals totals = usageService.getDailyTotals(tenant.getId(), java.time.LocalDate.now(java.time.ZoneOffset.UTC));
      double maxTokens = policy.getMaxTokensPerDay();
      double maxCost =
          policy.getMaxCostPerDayUsd() != null ? policy.getMaxCostPerDayUsd().doubleValue() : 0;

      if (maxTokens > 0) {
        double ratio = totals.tokens() / maxTokens;
        if (ratio >= 1) {
          alerts.add(new UsageAlert(tenant.getId(), "tokens", "critical", "Límite de tokens alcanzado", (double) totals.tokens(), maxTokens));
        } else if (ratio >= 0.8) {
          alerts.add(new UsageAlert(tenant.getId(), "tokens", "warning", "Tokens cerca del límite", (double) totals.tokens(), maxTokens));
        }
      }

      if (maxCost > 0) {
        double ratio = totals.costUsd() / maxCost;
        if (ratio >= 1) {
          alerts.add(new UsageAlert(tenant.getId(), "cost", "critical", "Límite de coste alcanzado", totals.costUsd(), maxCost));
        } else if (ratio >= 0.8) {
          alerts.add(new UsageAlert(tenant.getId(), "cost", "warning", "Coste cerca del límite", totals.costUsd(), maxCost));
        }
      }
    }

    return alerts;
  }

  public record UsageAlert(
      String tenantId, String type, String severity, String message, Double value, Double limit) {}
}

File: backend-javascript/src/usage/usage-alerts.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Policy } from '../common/entities/policy.entity';
import { Tenant } from '../common/entities/tenant.entity';
import { UsageService } from './usage.service';

export type UsageAlert = {
  tenantId: string;
  type: 'tokens' | 'cost' | 'kill_switch';
  severity: 'warning' | 'critical';
  message: string;
  value?: number;
  limit?: number;
};

@Injectable()
export class UsageAlertsService {
  constructor(
    @InjectRepository(Tenant)
    private readonly tenantsRepository: Repository<Tenant>,
    @InjectRepository(Policy)
    private readonly policiesRepository: Repository<Policy>,
    private readonly usageService: UsageService
  ) {}

  async list(tenantId?: string): Promise<UsageAlert[]> {
    let resolvedTenants: Tenant[] = [];
    if (tenantId) {
      const tenant = await this.tenantsRepository.findOne({ where: { id: tenantId } });
      resolvedTenants = tenant ? [tenant] : [];
    } else {
      resolvedTenants = await this.tenantsRepository.find();
    }

    const alerts: UsageAlert[] = [];

    for (const tenant of resolvedTenants as Tenant[]) {
      if (tenant.killSwitch) {
        alerts.push({
          tenantId: tenant.id,
          type: 'kill_switch',
          severity: 'critical',
          message: 'Kill switch activo'
        });
      }

      const policy = await this.policiesRepository.findOne({ where: { tenantId: tenant.id } });
      if (!policy) {
        continue;
      }

      const totals = await this.usageService.getDailyTotals(tenant.id);

      const maxTokensPerDay = Number(policy.maxTokensPerDay || 0);
      const maxCostPerDayUsd = Number(policy.maxCostPerDayUsd || 0);

      if (maxTokensPerDay > 0) {
        const ratio = totals.tokens / maxTokensPerDay;
        if (ratio >= 1) {
          alerts.push({
            tenantId: tenant.id,
            type: 'tokens',
            severity: 'critical',
            message: 'Límite de tokens alcanzado',
            value: totals.tokens,
            limit: maxTokensPerDay
          });
        } else if (ratio >= 0.8) {
          alerts.push({
            tenantId: tenant.id,
            type: 'tokens',
            severity: 'warning',
            message: 'Tokens cerca del límite',
            value: totals.tokens,
            limit: maxTokensPerDay
          });
        }
      }

      if (maxCostPerDayUsd > 0) {
        const ratio = totals.costUsd / maxCostPerDayUsd;
        if (ratio >= 1) {
          alerts.push({
            tenantId: tenant.id,
            type: 'cost',
            severity: 'critical',
            message: 'Límite de coste alcanzado',
            value: totals.costUsd,
            limit: maxCostPerDayUsd
          });
        } else if (ratio >= 0.8) {
          alerts.push({
            tenantId: tenant.id,
            type: 'cost',
            severity: 'warning',
            message: 'Coste cerca del límite',
            value: totals.costUsd,
            limit: maxCostPerDayUsd
          });
        }
      }
    }

    return alerts;
  }
}
